[SYSTEMDS-3546] Push down image pre-processing to blas/mkl

anishsapkota opened the pull request:
This is where i intend to add the native bindings for intel MKL image processing.

anishsapkota commented on src/main/cpp/libAnish.cpp:
Hi Sebastian, 

i have successfully implemented a mkl degmm for test purpose . For that i did the native function implementation in systemds.cpp, function defination in systemds.h and created a native interface in NativeHelper.java. In src/test/java/org/apache/sysds/test/functions i have also implemented a main function in Java to test the functionality. it works with mkl and succefully calls the native cpp function. 

However, i had to comment out the conflicting functions of openBLAS in order to build the shared lib. :(
i will need your help to find out another way around.

Baunsgaard replied:
for the future, if you have general comments i suggest you put them in the general comments, not as a comment on a comment like this. Thanks.

Baunsgaard commented on src/main/java/org/apache/sysds/utils/NativeHelper.java:
```diff
@@ -416,4 +416,7 @@ public static native boolean conv2dSparse(int apos, int alen, int[] aix, double[
 	// fastest. We can revisit this decision later and hence I would not recommend removing this method.
 	private static native void setMaxNumThreads(int numThreads);
+
+	//test interface for native binding
```
it is not a test inteface anymore, 
please write a full Javadoc definition using:

/** */

Baunsgaard commented on src/test/java/org/apache/sysds/test/functions/native/NativeBindingTestWithDgemm.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Random;
```
license

Baunsgaard commented on src/test/java/org/apache/sysds/test/functions/native/NativeBindingTestWithDgemm.java:
```diff
@@ -0,0 +1,74 @@
+            for (int j = 0; j < n; j++) {
+                System.out.print(flatC[i * n + j] + " ");
+            }
+            System.out.println();
```
avoid printing in tests. 
A general design for clean tests is to only print once something goes wrong. 
you can imagine if we have a million tests that print. then we would not be able to decipher what is wrong or what is right. (and we would use all our time printing in the tests)

Baunsgaard commented on src/test/java/org/apache/sysds/test/functions/native/NativeBindingTestWithDgemm.java:
```diff
@@ -0,0 +1,74 @@
+        }
+    }
+
+    public static void main(String[] args) {
```
you can leave this function for tests purposes, but ideally please integrate the test you make into JUnit.
this is as simple as marking methods with \@ Test  See other files for examples.

Baunsgaard reviewed (commented):
This is nice progress, 


I understand based on your comment that commenting out other code this was the only way for you to make it work.

But we should strive to find a way to make all of it work (your point as well)
perhaps the limitation is that you do not have BLAS installed as well, and maybe there is a requirement that you need to install both to compile the library?

Alternatively we can add a new cpp file called systemds_img.cpp and header file that you hook into. But then you need to make a new NativeHelper class instance that then hooks up to the new file. In general i would be interested if there was a nice way of doing such a thing.

Thanks

Baunsgaard commented:
Hi @anishsapkota,

How is it going? is there uncommitted progress or do you need any further comments?
Also, feel free if any of the comments are resolved, to click resolved yourself.

anishsapkota commented:
Hi @Baunsgaard ,
I have been loading the shared libraries manually on my tests as the Native Helper currently supports LINUX and Windows. I have been working using MAC_OS and the libraries are therefore not loaded automatically. Could you please suggest me something on this?

Apart from this, I have created a new file systemds_img.cpp and add the methods that I have implemented there, added nativ function tests to JavaTest.yaml and created a new Java Class ImgNativeHelper that extends NativeHelper.

Baunsgaard commented on src/test/java/org/apache/sysds/test/functions/nativ/ImgUtilsTest.java:
```diff
@@ -0,0 +1,412 @@
+public class ImgUtilsTest {
+
+    /*static {
+        System.loadLibrary("systemds_openblas-Darwin-x86_64");
```
I would simply add such a static block inside the ImgNativeHelper.java that loads the correctly compiled file depending on the OS in use, if there is no such file error out.
The reason the Native loader does not work on Mac is simply because we have not yet compiled the files for this.

anishsapkota replied:
Thank you, 
while adding the static block, can I assume that the java.library.path is already configured ?

anishsapkota replied:
but still, we need to compile the new implementations in order to include them into the native library. Or should I push my own compiled lib for test ?

Baunsgaard replied:
No we compile it on each test run for that machine. You can find the appropriate code in our GitHub action script.

anishsapkota commented:
Hi @Baunsgaard ,
I am waiting for the tests to run, so that we can merge.

Baunsgaard commented:
> Hi @Baunsgaard , I am waiting for the tests to run, so that we can merge.
> 
> regarding the performance results, should I be comparing the performance between OpenBlas and Intel MKL ? or with the existing DML implementation?

You need to compare the results between all possible ones you have.
But most importantly between DML Java version and your MKL.

Baunsgaard commented:
Regarding the tests, I need to allow them to run since this is your first commit it does not allow them to automatically run.
I will keep an eye on it and run the tests when you commit, if you feel like you are waiting to long just write a comment.

anishsapkota commented:
java.lang.UnsatisfiedLinkError: /github/workspace/src/main/cpp/lib/libsystemds_openblas-Linux-x86_64.so: libopenblas.so.0: cannot open shared object file: No such file or directory

I still can't figure it out why it can't find the lib. could you please review the static block from my last commit ?

Baunsgaard commented:
We call this script in the beginning of building
[
docker/entrypoint.sh](https://github.com/apache/systemds/blob/74be3fcdba8ec1015b860ce4f5a9a6758c306ac9/docker/entrypoint.sh)

can it be that:

```
cd /github/workspace/src/main/cpp
./build.sh
cd /github/workspace
```

does not build your binaries?

Baunsgaard commented:
> java.lang.UnsatisfiedLinkError: /github/workspace/src/main/cpp/lib/libsystemds_openblas-Linux-x86_64.so: libopenblas.so.0: cannot open shared object file: No such file or directory
> 
> I still can't figure it out why it can't find the lib. could you please review the static block from my last commit ?

it does not look wrong in the last commit.

anishsapkota commented:
> We call this script in the beginning of building [ docker/entrypoint.sh](https://github.com/apache/systemds/blob/74be3fcdba8ec1015b860ce4f5a9a6758c306ac9/docker/entrypoint.sh)
> 
> can it be that:
> 
> ```
> cd /github/workspace/src/main/cpp
> ./build.sh
> cd /github/workspace
> ```
> 
> does not build your binaries?

how can I verify that? 
there were old binaries in the lib folder. When the git tests run, it should build new ones and replace the old one with it right ?

even if this doesn't happen, the System.load() should load the old binaries that are present in cpp/lib and throw some error that says it could not find the native implementations.

Baunsgaard commented:
> > We call this script in the beginning of building [ docker/entrypoint.sh](https://github.com/apache/systemds/blob/74be3fcdba8ec1015b860ce4f5a9a6758c306ac9/docker/entrypoint.sh)
> > can it be that:
> > ```
> > cd /github/workspace/src/main/cpp
> > ./build.sh
> > cd /github/workspace
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > does not build your binaries?
> 
> how can I verify that? there were old binaries in the lib folder. When the git tests run, it should build new ones and replace the old one with it right ?
> 
> even if this doesn't happen, the System.load() should load the old binaries that are present in cpp/lib and throw some error that says it could not find the native implementations.

we always build the binaries on the GitHub testing.

you can look up the raw logs :

https://pipelinesghubeus22.actions.githubusercontent.com/99y4vYqbkvSsdnIzLIwwBWoa7RjP8bOodIdS7qcJ1UCIIzC9x5/_apis/pipelines/1/runs/40826/signedlogcontent/30?urlExpires=2023-09-11T15%3A56%3A29.5229959Z&urlSigningMethod=HMACV1&urlSignature=O1AR7JTkHl6vk70WkmpWS8AHt98TXruku%2B9w0t14eJI%3D

and there it says something like:

```
2023-09-11T13:45:53.2908375Z See also "/github/workspace/src/main/cpp/OPENBLAS/CMakeFiles/CMakeOutput.log".
2023-09-11T13:45:53.3071180Z make: *** No rule to make target 'install'.  Stop.
2023-09-11T13:45:53.3123270Z -----------------------------------------------------------------------
2023-09-11T13:45:53.3123866Z Check for unexpected dependencies added after code change or new setup:
2023-09-11T13:45:53.3124457Z Non-standard dependencies for libsystemds_mkl-linux-x86_64.so
2023-09-11T13:45:53.3192053Z Non-standard dependencies for libsystemds_openblas-linux-x86_64.so
2023-09-11T13:45:53.3264477Z -----------------------------------------------------------------------
2023-09-11T13:45:53.4061848Z -- The CXX compiler identification is GNU 9.4.0
2023-09-11T13:45:53.4118015Z -- Check for working CXX compiler: /usr/bin/g++
2023-09-11T13:45:53.5109326Z -- Check for working CXX compiler: /usr/bin/g++ -- works
2023-09-11T13:45:53.5126103Z -- Detecting CXX compiler ABI info
2023-09-11T13:45:53.6161549Z -- Detecting CXX compiler ABI info - done
2023-09-11T13:45:53.6300600Z -- Detecting CXX compile features
2023-09-11T13:45:53.6301203Z -- Detecting CXX compile features - done
2023-09-11T13:45:53.6334317Z -- Looking for C++ include pthread.h
2023-09-11T13:45:53.7454940Z -- Looking for C++ include pthread.h - found
2023-09-11T13:45:53.7455614Z -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
2023-09-11T13:45:53.8619331Z -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
2023-09-11T13:45:53.8621643Z -- Check if compiler accepts -pthread
2023-09-11T13:45:53.9742424Z -- Check if compiler accepts -pthread - yes
2023-09-11T13:45:53.9759399Z -- Found Threads: TRUE  
2023-09-11T13:45:53.9771762Z -- Microsoft SEAL -> Version 3.7.0 detected
2023-09-11T13:45:53.9789595Z -- Microsoft SEAL -> Targets available: SEAL::seal_shared
2023-09-11T13:45:53.9790366Z -- Configuring done
2023-09-11T13:45:53.9811958Z -- Generating done
2023-09-11T13:45:53.9817535Z -- Build files have been written to: /github/workspace/src/main/cpp/HE
2023-09-11T13:45:54.0510546Z Scanning dependencies of target he
2023-09-11T13:45:54.0700122Z [ 33%] Building CXX object CMakeFiles/he.dir/libhe.cpp.o
2023-09-11T13:45:58.4275559Z [ 66%] Building CXX object CMakeFiles/he.dir/he.cpp.o
2023-09-11T13:46:02.6181043Z [100%] Linking CXX shared library libhe-Linux-x86_64.so
2023-09-11T13:46:02.7196534Z [100%] Built target he
2023-09-11T13:46:02.7462105Z Install the project...
```

and that indicate to me that the image lib is not compiled.

we need to add it as an element to compile in: 

https://github.com/apache/systemds/blob/74be3fcdba8ec1015b860ce4f5a9a6758c306ac9/src/main/cpp/build.sh

Baunsgaard commented:
> <img alt="peformance_eval" width="604" src="https://user-images.githubusercontent.com/68165368/267070022-2f5105d5-36dc-4ce4-bbc4-eab517940105.png">
> 
> here is a result from the performance evaluation on my local machine. I am little bit unsure about the intel mkl, i used this by setting MKL_NUM_THREADS = 4 as I found that it works best for my machine. It was little bit surprising that although I have intel chip, OpenBlas significantly performed better.

This already looks promising, 

please run the tests with repeats such that you get the average run time of ~100 runs.
also run it both with a sparse input and a dense input.

Thanks!

anishsapkota commented:
> and there it says something like:

yes, I actually adjusted it in CMakeLists.txt, it includes the img lib and the build.sh script should use CMakeList.txt to compile those binaries right ?

Baunsgaard commented:
> > and there it says something like:
> 
> yes, I actually adjusted it in CMakeLists.txt, it includes the img lib and the build.sh script should use CMakeList.txt to compile those binaries right ?
> 
> CMakeLists.txt set(HEADER_FILES imgUtils.h libmatrixdnn.h libmatrixmult.h systemds.h common.h) set(SOURCE_FILES imgUtils.cpp systemds_img.cpp libmatrixdnn.cpp libmatrixmult.cpp systemds.cpp)

It should yes, but maybe there is something else going wrong. 
Perhaps adding another block of code in the GitHub action. 

```txt
    - name: Build C++
      run: |
        cd /github/workspace/src/main/cpp
        ./build.sh
        ls lib
```

you can add it after the checkout block.
Maybe that gives you some indications?

anishsapkota commented:
Performance Evaluation averaged over 100 runs:
<img width="518" alt="Screenshot 2023-09-11 at 19 59 44" src="https://github.com/apache/systemds/assets/68165368/797db0d7-1c44-4475-8d19-94efa751b152">

I observed that the first run takes significantly more time in comparison to later runs, although the input matrices are completely randomised.

Baunsgaard commented:
> Performance Evaluation averaged over 100 runs: <img alt="Screenshot 2023-09-11 at 19 59 44" width="518" src="https://user-images.githubusercontent.com/68165368/267107362-797db0d7-1c44-4475-8d19-94efa751b152.png">
> 
> I observed that the first run takes significantly more time in comparison to later runs, although the input matrices are completely randomised.

great! 

There is an option of removing outliers what I typically do is remove the 5% best and worst performing runs and then do the average of the remaining. Also good practice is to calculate std on the runs and mark it in the graphs with +-, this is similarly done without the 5%.

Also for more clarification of where time is spend I would love it if you run SystemDS with the -stats  flag.
this can indicate where we are using time in the DML implementation

anishsapkota commented:
/Users/runner/work/_temp/b8ad4948-f59d-4334-ad56-0415e504b59f.sh: line 1: cd: /github/workspace/src/main/cpp: No such file or directory

error during the build after the latest commit

Baunsgaard commented:
> /Users/runner/work/_temp/b8ad4948-f59d-4334-ad56-0415e504b59f.sh: line 1: cd: /github/workspace/src/main/cpp: No such file or directory
> 
> error during the build after the latest commit

okay, then since it does not work... lets do relative paths like in:

```
    - name: Build Python Package
      run: |
        cd src/main/python
        python create_python_dist.py
```

Baunsgaard commented:
Looking at the other PR, i remembered,

Please also test with uneven images with different number of row and column pixels.

Baunsgaard commented:
https://github.com/apache/systemds/actions/runs/6151101437/job/16690515871?pr=1843

Some of the errors are to be expected Since some packages are missing, but it seems clear to me that it does not build the img packages. Could it be that we are missing the MKL and BLAS dependencies on the GitHub install?

if so is there some install sequence that can be run on Ubuntu, then we can make it work at least for Linux.

anishsapkota commented:
> Looking at the other PR, i remembered,
> 
> Please also test with uneven images with different number of row and column pixels.

I will try to incorporate this as well. Currently only the square images are supported, at least for rotation.
if I try to rotate an image lets say 4 x 3 by 90 degrees, then the output image has to be 3 x 4 and vice versa? And what if I were to rotate it by 45 degrees, I assume the dimension remains the same but certain part of image will be cut off.

Baunsgaard commented:
> > Looking at the other PR, i remembered,
> > Please also test with uneven images with different number of row and column pixels.
> 
> I will try to incorporate this as well. Currently only the square images are supported, at least for rotation. if I try to rotate an image lets say 4 x 3 by 90 degrees, then the output image has to be 3 x 4 and vice versa? And what if I were to rotate it by 45 degrees, I assume the dimension remains the same but certain part of image will be cut off.

it is okay if some things are not supported, in such cases we just error out with a reasonable message.

anishsapkota commented:
Hi @Baunsgaard , 
could you please let the tests run, once you are online. Thank you.

anishsapkota commented:
Hi @Baunsgaard 
The build was successful for Linux using openblas, little progress. 
However I noticed that the built shared library location is different from the workspace.
/home/runner/work/systemds/systemds/src/main/cpp/lib/libsystemds_openblas-Linux-x86_64.so

Baunsgaard commented:
> Hi @Baunsgaard The build was successful for Linux using openblas, little progress. However I noticed that the built shared library location is different from the workspace. /home/runner/work/systemds/systemds/src/main/cpp/lib/libsystemds_openblas-Linux-x86_64.so
> 
> I am trying again with this absolute path.

figuring out where all these things are, can be tricky. hope it works out it is a bit of trial and error.

anishsapkota commented:
I was going through the failed javaTest, and apparently I am installing the openblas outside of the docker container, I put the install section in build.yml file https://github.com/apache/systemds/pull/1843/commits/c500e1192665f3bea8df0a88135b62857635add2

should I be putting install OpenBlas section in action.yml ?

-- Configuring incomplete, errors occurred!
See also "/github/workspace/src/main/cpp/INTEL/CMakeFiles/CMakeOutput.log".
make: *** No rule to make target 'install'.  Stop.
-- The CXX compiler identification is GNU 9.4.0
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at cmake/FindOpenBLAS.cmake:72 (MESSAGE):
-- Could not find OpenBLAS lib. Turning OpenBLAS_FOUND off
  Could not find OpenBLAS
Call Stack (most recent call first):
  CMakeLists.txt:69 (find_package)

Baunsgaard commented:
> I was going through the failed javaTest, and apparently I am installing the openblas outside of the docker container, I put the install section in build.yml file [c500e11](https://github.com/apache/systemds/commit/c500e1192665f3bea8df0a88135b62857635add2)
> 
> should I be putting install OpenBlas section in entrypoint.sh ?
> 
> -- Configuring incomplete, errors occurred! See also "/github/workspace/src/main/cpp/INTEL/CMakeFiles/CMakeOutput.log". make: *** No rule to make target 'install'. Stop. -- The CXX compiler identification is GNU 9.4.0 -- Check for working CXX compiler: /usr/bin/g++ -- Check for working CXX compiler: /usr/bin/g++ -- works -- Detecting CXX compiler ABI info -- Detecting CXX compiler ABI info - done -- Detecting CXX compile features -- Detecting CXX compile features - done CMake Error at cmake/FindOpenBLAS.cmake:72 (MESSAGE): -- Could not find OpenBLAS lib. Turning OpenBLAS_FOUND off Could not find OpenBLAS Call Stack (most recent call first): CMakeLists.txt:69 (find_package)

You added it in Build. therefore yes it is not in the JavaTest GitHub Action. 
I thought you just tried some things out in build, to see what to do to make it work.

I would suggest to modify the actual build.sh script in src/main/cpp 
that is running every time even inside the docker image.

in that build.sh script you can detect if the correct requirements are installed and if not install them.

anishsapkota commented:
Hi @Baunsgaard 
I think there is something wrong with my last commit, not even a single javaTest ran.  is it possible to intervene this and run the latest commit, the one with "OneApi"?

anishsapkota commented:
@Baunsgaard could you please approved the workflow again, previous installation got stuck as it was waiting for user interaction.

anishsapkota commented:
Hi @Baunsgaard , 
could you please review the javaTest that are failing?  They are not the part of my implementations and seem to somehow fail, I could not figure it out why

anishsapkota commented:
Hi @Baunsgaard, 
I have a situation here. I was able to successfully install INTEL oneApi Basekit which includes newer Intel MKL lib earlier and run the setvars.sh script to set the env variables. But the installation conflicts with the earlier implementations made with older INTEL MKL. It says "mkl_dnn.h" not found during build. 

The older version of intel from 2020 also does not includes "mkl_dnn.h" either. 

Please guide me through this.

j143 commented:
Hi @anishsapkota -- Now, the tests are running.

anishsapkota commented:
Hi @j143 ,

At the moment https://github.com/apache/systemds/pull/1843/commits/779131a24d26d8f210fc77340284a39d85deaae7, I am really confused that the shared object libsystemds_mkl-Linux-x86_64.so is not being found under 
/github/workspace/src/main/cpp/lib although the other shared object libsystemds_openblas-Linux-x86_64.so is being loaded from same path and the tests run through.

Moreover random tests are failing, like countDistinct, built-in.part2, which would otherwise pass, if I were not to install intel oneAPi. 

The biggest problem is the header file "mkl_dnn.h" which is imported into libmatrixdnn.cpp, it doesn't exist in Intel oneApi include folder. 
https://github.com/apache/systemds/blob/main/src/main/cpp/libmatrixdnn.cpp

https://github.com/01org/mkl-dnn/releases/download/v0.10/mklml_lnx_2018.0.20170908.tgz
This release contains the "mkl_dnn.h"

should I try to unzip it and add it somewhere and export its path to CPATH ?

j143 commented:
> I am really confused that the shared object libsystemds_mkl-Linux-x86_64.so is not being found under
/github/workspace/src/main/cpp/lib although the other shared object libsystemds_openblas-Linux-x86_64.so is being loaded from same path and the tests run through.

you check the github actions workflow logs to find out why this is the case.

> Moreover random tests are failing, like countDistinct, built-in.part2, which would otherwise pass, if I were not to install intel oneAPi.

this can be taken once the above is resolved, I guess.

> should I try to unzip it and add it somewhere and export its path to CPATH ?

The dnn library seems to be published differently. You could choose to download and add it. (we can't commit mkl_dnn.h to the systemds code btw.)


I am running this workflow once again, so that you could get the logs to debug.

anishsapkota commented:
libhe-Linux-x86_64.so
libsystemds_mkl-Linux-x86_64.so
libsystemds_mkl-Windows-AMD64.dll
libsystemds_openblas-Linux-x86_64.so
libsystemds_openblas-Windows-AMD64.dll
libsystemds_spoof_cuda-Linux-x86_64.so
libsystemds_spoof_cuda-Windows-AMD64.dll

this was the output of ls /github/workspace/src/main/cpp/lib **libsystemds_mkl-Linux-x86_64.so** seems to be there.

Error:  org.apache.sysds.test.functions.nativ.ImgUtilsTest.runTestsWithMKL  Time elapsed: 0.001 s  <<< ERROR!
**java.lang.UnsatisfiedLinkError: /github/workspace/src/main/cpp/lib/libsystemds_mkl-Linux-x86_64.so:** libmkl_rt.so: cannot open shared object file: **No such file or directory**
	at org.apache.sysds.test.functions.nativ.ImgUtilsTest.runTests(ImgUtilsTest.java:497)
	at org.apache.sysds.test.functions.nativ.ImgUtilsTest.runTestsWithMKL(ImgUtilsTest.java:35)
it doesn't make sense, 

**libsystemds_openblas-Linux-x86_64.so** this one is however loaded without any problem.

Baunsgaard commented:
> libhe-Linux-x86_64.so libsystemds_mkl-Linux-x86_64.so libsystemds_mkl-Windows-AMD64.dll libsystemds_openblas-Linux-x86_64.so libsystemds_openblas-Windows-AMD64.dll libsystemds_spoof_cuda-Linux-x86_64.so libsystemds_spoof_cuda-Windows-AMD64.dll
> 
> this was the output of ls /github/workspace/src/main/cpp/lib **libsystemds_mkl-Linux-x86_64.so** seems to be there.
> 
> Error: org.apache.sysds.test.functions.nativ.ImgUtilsTest.runTestsWithMKL Time elapsed: 0.001 s <<< ERROR! **java.lang.UnsatisfiedLinkError: /github/workspace/src/main/cpp/lib/libsystemds_mkl-Linux-x86_64.so:** libmkl_rt.so: cannot open shared object file: **No such file or directory** at org.apache.sysds.test.functions.nativ.ImgUtilsTest.runTests(ImgUtilsTest.java:497) at org.apache.sysds.test.functions.nativ.ImgUtilsTest.runTestsWithMKL(ImgUtilsTest.java:35)
> 
> it doesn't make sense,
> 
> **libsystemds_openblas-Linux-x86_64.so** this one is however loaded without any problem from same path

i am unfortunately unaware what needs to be loaded when, there must be some sequence or elements missing. this libmkl_rt.so is what is says is missing now, not your build mkl file.


About unrelated tests to native, none of them should be effected by your current changes, so if they fail it is okay, we can rerun them to see if it is a systematic error. On a side node i suspect that they run the PR actions on some other hardware that sometimes is more prone to errors, but i have no way of confirming that.

anishsapkota commented:
> i am unfortunately unaware what needs to be loaded when, there must be some sequence or elements missing. this libmkl_rt.so is what is says is missing now, not your build mkl file.

I was misunderstanding the log output, I have added LD_LIBRARY_PATH and additional ls statements to see the outputs, could you please lt it run once again, @Baunsgaard

anishsapkota commented:
@Baunsgaard upon checking the env variables, the LD_LIBRARY_PATH and LIBRARY_PATH both contains the path to the folder where libmkl_rt.so is.

anishsapkota commented:
**-- Found MKL (include: /opt/intel/oneapi/mkl/2023.2.0/include, lib: "/opt/intel/oneapi/mkl/2023.2.0/lib/intel64/libmkl_rt.so"**

it finds the shared object during the build. But I can't figure it out why it's causing problem during loading. As I don't have a linux system I really cannot further investigate it and solely rely on these tests which is very inefficient. I tried using the UNI's Linux terminals, but they do not allow any package installations from the internet. My laptop would crash while using Virtual box with Ubuntu. :(

anishsapkota commented:
@Baunsgaard @j143 , could you please let the last commit run?

Baunsgaard commented:
maybe the tests run on your fork? 
you can disable some of them to not run all the time and then run it there.
note your are limited to 5 at a time so i was trying to avoid this.

to do this, 
1. comment out tests that are not relevant for you such as build license python and most of the java tests
3. open the PR on your fork as a PR into your main branch (on your fork.).

j143 commented:
Hi @anishsapkota  -- you can go to actions tab in your personal fork (ex, https://github.com/anishsapkota/systemds/actions)  and enable the workflows! you shall be able to run these tests there too! you can disable workflows that are not related to your changes.

also, you can [enable debug option](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging), if required to get even more logs for you test case.

anishsapkota commented:
Thank you @Baunsgaard  @j143 , 
Now I can run the tests on my own. :)

j143 commented:
Hi @Baunsgaard 

> System.IO.IOException: No space left on device : '/home/runner/runners/2.309.0/_diag/Worker_20230915-060835-utc.log'
Unhandled exception. System.IO.IOException: No space left on device : '/home/runner/runners/2.309.0/_diag/Worker_20230915-060835-utc.log'

https://github.com/apache/systemds/actions/runs/6191936104

Baunsgaard commented:
> Hi @Baunsgaard
> 
> > System.IO.IOException: No space left on device : '/home/runner/runners/2.309.0/_diag/Worker_20230915-060835-utc.log'
> > Unhandled exception. System.IO.IOException: No space left on device : '/home/runner/runners/2.309.0/_diag/Worker_20230915-060835-utc.log'
> 
> https://github.com/apache/systemds/actions/runs/6191936104

well, i guess we need to cleanup some of the resources. 
I think we only have 8GB disk space, but i might remember it wrong. 

what we need to do is to remove cached files from install of our docker image to reduce its size via `apt-get clean`

docker/testsysds.Dockerfile

Then include MKL and OpenBLAS  in the image rather than installing them in the build script of cpp and make sure we do not keep around anything not needed in our install.

Baunsgaard commented:
> @Baunsgaard @j143 Hi, after some intensive trying using **gitpod** I could conclude following things:
> 
>     1. **libmkl_rt.so cannot open the file: no such file or directory** this error was because ldconfig could not find the exported LD_LIBRARY_PATH. I could solve this problem by adding a **.conf file** with **/opt/intel/oneapi/mkl/latest/lib/intel64** in **/etc/ld.so.conf.d** directory and reloading **ldconfig.**
> 
>     2. Next Thanks to @j143 I stumbled across his video about Intel MKL installation on ubuntu. If I would do it just like he mentioned using a **offline package installer** which has CLI based UI and requires user interaction ( _I could not find a way to do this using shell script_), everything would work fine after doing what I mentioned in 1.
>        If we were to use **apt package installer** instead of offline one, somehow the include folder containing all the necessary header file like "mkl.h" is missing in mkl root folder. I tried using "locate mkl.h" no result.
>        I could not find the solution to this.
>        Next option would be to install whole intel-basekit which contains MKL as well. It is unfortunately too large ~13 GB I think and could have led to the not enough space error.
> 
>     3. Regarding "mkl_dnn.h" .I have mentioned here [[SYSTEMDS-3546] Push down image pre-processing to blas/mkl #1843 (comment)](https://github.com/apache/systemds/pull/1843#issuecomment-1718879255), if we download it unzip it, we have 2 folders in it include and lib, then copy "mkl_dnn.h" and "mkl_dnn_types.h"(second name is something similar) for its include folder to /opt/intel/oneapi/mkl/latest/include (if present) and the libs to /opt/intel/oneapi/mkl/latest/lib/intel64. Using this trick I could compile the cpp files with "mkl_dnn.h"header file as well. **But not recommended it was just an experiment.**
> 
>     4. Nevertheless, I could run the benchmarks using both mkl and openblas, also DML implementations using -stats flag in gitpod. Although I could not install mkl properly in GitHub test docker container. I will process the results and post it.
> 
> 
> @Baunsgaard should I comment out "mkl test case" so that we can merge? or how should I proceed now?

lets go with the Blas test then and add some TODO that reference this PR about the MKL tests, then I will try to make it work at a later time, since i need it for other things.

Baunsgaard commented on .github/workflows/javaTests.yml:
```diff
@@ -81,39 +82,39 @@ jobs:
-        path: target/jacoco.exec
-        retention-days: 1
+      - name: Checkout Repository
+        uses: actions/checkout@v4
```
we need to revert here, since the change is only syntax.

Baunsgaard commented on src/main/cpp/build.sh:
```diff
@@ -46,6 +47,14 @@ if ! [ -x "$(command -v patchelf)" ]; then
+
+  apt-get update
+  apt-get install libopenblas-dev -y
+fi
```
add comment on MKL install.

Baunsgaard commented on src/main/cpp/imgUtils.cpp:
```diff
@@ -0,0 +1,243 @@
+        cout <<  endl;
+    }
+    cout << "\n"<< endl;
+}
```
i suggest removing this debugging printImage.

Baunsgaard commented on src/main/cpp/imgUtils.cpp:
```diff
@@ -0,0 +1,243 @@
+               cblas_dcopy(end_x - start_x, &img_in[(x_in) + (y_in) * in_w],
+                           1, &img_out[y * out_w + start_x + static_cast<int>(offset_x)], 1);
+           }
+       }
```
indentation.

Baunsgaard commented on src/main/cpp/imgUtils.cpp:
```diff
@@ -0,0 +1,243 @@
+    return img_out;
+}
+
+void img_translate(double* img_in, double offset_x, double offset_y,
```
cast the offsets in the first line rather than every call to offset_x and y, and maybe consider changing the API to use int offsets.

Baunsgaard commented on src/main/cpp/imgUtils.cpp:
```diff
@@ -0,0 +1,243 @@
+        }
+
+        // Perform matrix multiplication to compute sampling pixel indices
+        double* transformed_coords = new double[2 * out_w * out_h];
```
i do not know, but 

can this be allocated as int?

Baunsgaard commented on src/main/cpp/imgUtils.h:
```diff
@@ -0,0 +1,38 @@
+#define IMAGE_UTILS_H
+#include <cmath>
+
+void printImage(const double* image, int rows, int cols);
```
consider removing

Baunsgaard commented on src/main/cpp/lib/libsystemds_mkl-Linux-x86_64.so:
maybe not add these?

Baunsgaard commented on src/main/cpp/systemds_img.cpp:
```diff
@@ -0,0 +1,135 @@
+    delete[] img_out;
+
+    return img_out_java;
+}*/
```
remove commented code.

Baunsgaard commented on src/main/cpp/systemds_img.cpp:
```diff
@@ -0,0 +1,135 @@
+    img_translate(j_img_in, offset_x, offset_y, in_w, in_h, out_w, out_h, fill_value, j_img_out);
+    env->ReleaseDoubleArrayElements(img_in, j_img_in, 0);
+    env->ReleaseDoubleArrayElements(img_out, j_img_out, 0);
+}
```
newline in the end.

Baunsgaard commented on src/main/java/org/apache/sysds/utils/ImgNativeHelper.java:
```diff
@@ -0,0 +1,112 @@
+                System.load("/src/main/cpp/lib/".replace("/", File.separator) + "libsystemds_" + libname);
+            } else if (SystemUtils.IS_OS_MAC || SystemUtils.IS_OS_MAC_OSX) {
+               // String libname = "systemds_" + blasType + "-Darwin-x86_64";
+                System.load(System.getProperty("user.dir")+"/src/main/cpp/lib/libsystemds_" + blasType + "-Darwin-x86_64.dylib");
```
same here, remove commented code,

anishsapkota commented:
> My main comment is the comparison is not fair. There are some things we can do to improve it if you want (read the detailed comments for more).

Hi @Baunsgaard ,
thank your for reviewing the code and the suggestions. Rest assured, I will address all of them after my presentation.

could you be more specific, how can I make the comparison fairer? I will be incorporating your previous suggestions of calculating std and analysing dml execution time using -stats flag. Do you have any further suggestions?
It would be interesting for the presentation to address this issue before that.

Baunsgaard commented:
> > My main comment is the comparison is not fair. There are some things we can do to improve it if you want (read the detailed comments for more).
> 
> Hi @Baunsgaard , thank your for reviewing the code and the suggestions. Rest assured, I will address all of them after my presentation.
> 
> could you be more specific, how can I make the comparison fairer? I will be incorporating your previous suggestions of calculating std and analysing dml execution time using -stats flag. Do you have any further suggestions? It would be interesting for the presentation to address this issue before that.

The fairest comparison would be to integrate the native instruction as a DML supported command. It does not take long to make. But requires you to know where to do it. 
Once integrated we can compare with the overhead of the system on both executions and thereby get a fair comparison. 

For now the stats comparison is close enough.

anishsapkota commented:
> The fairest comparison would be to integrate the native instruction as a DML supported command. It does not take long to make. But requires you to know where to do it.

could you please let me know which files I should be looking at, so that I can implement this as well ?

anishsapkota commented:
<img width="771" alt="image" src="https://github.com/apache/systemds/assets/68165368/5c2c8459-2407-4dce-ab09-c6360fd13d70">

anishsapkota commented:
<img width="761" alt="image" src="https://github.com/apache/systemds/assets/68165368/29e18030-3840-4292-aa02-b4fab4797af7">

anishsapkota commented:
<img width="793" alt="image" src="https://github.com/apache/systemds/assets/68165368/c656f62c-1f6c-44db-acfc-c4a65135881b">

anishsapkota commented:
<img width="600" alt="image" src="https://github.com/apache/systemds/assets/68165368/9236727e-a9fe-4b8f-89b7-3bd89c7fe367">

anishsapkota commented:
**image_rotate**

SystemDS Statistics:
Total elapsed time:		1,842 sec.
Total compilation time:		0,042 sec.
**Total execution time:		1,800 sec.**
Number of compiled Spark inst:	1.
Number of executed Spark inst:	0.
Cache hits (Mem/Li/WB/FS/HDFS):	1835090/0/0/0/1.
Cache writes (Li/WB/FS/HDFS):	3/14/0/1.
Cache times (ACQr/m, RLS, EXP):	0,946/0,352/1,060/0,058 sec.
HOP DAGs recompiled (PRED, SB):	0/0.
HOP DAGs recompile time:	0,167 sec.
Spark ctx create time (lazy):	0,000 sec.
Spark trans counts (par,bc,col):0/0/0.
Spark trans times (par,bc,col):	0,000/0,000/0,000 secs.
Spark async. count (pf,bc,op): 	0/0/0.
ParFor loops optimized:		1.
ParFor optimize time:		0,001 sec.
ParFor initialize time:		0,004 sec.
ParFor result merge time:	0,028 sec.
ParFor total update in-place:	0/262144/262168
Total JIT compile time:		36.02 sec.
Total JVM GC count:		175.
Total JVM GC time:		0.374 sec.
Heavy hitter instructions:
  #  Instruction      Time(s)    Count
  1  rightIndex         3,288   786443
  **2  m_img_rotate       1,711        1**
  **3  m_img_transform    1,711        1**
  4  rmvar              1,434  2359350
  5  leftIndex          1,213   262168
  6  createvar          0,997  1048631
  7  castdts            0,846   524294
  8  &&                 0,544   786432
  9  mvvar              0,392   524300
 10  <                  0,355   524288

**image_cutout**

SystemDS Statistics:
Total elapsed time:		0,108 sec.
Total compilation time:		0,042 sec.
**Total execution time:		0,065 sec.**
Number of compiled Spark inst:	1.
Number of executed Spark inst:	0.
Cache hits (Mem/Li/WB/FS/HDFS):	2/0/0/0/1.
Cache writes (Li/WB/FS/HDFS):	0/1/0/1.
Cache times (ACQr/m, RLS, EXP):	0,032/0,000/0,000/0,033 sec.
HOP DAGs recompiled (PRED, SB):	0/0.
HOP DAGs recompile time:	0,000 sec.
Spark ctx create time (lazy):	0,000 sec.
Spark trans counts (par,bc,col):0/0/0.
Spark trans times (par,bc,col):	0,000/0,000/0,000 secs.
Spark async. count (pf,bc,op): 	0/0/0.
Total JIT compile time:		13.577 sec.
Total JVM GC count:		176.
Total JVM GC time:		0.368 sec.
Heavy hitter instructions:
  #  Instruction  Time(s)  Count
  1  write          0,033      1
  **2  sp_rblk        0,032      1**
  3  rmvar          0,001      2
  4  leftIndex      0,000      1
  5  rand           0,000      1
  6  createvar      0,000      4
  7  +              0,000      2
  8  <              0,000      1
  9  mvvar          0,000      8
 10  -              0,000      2

**image_crop**

New w/h=409.0/409.0
SystemDS Statistics:
Total elapsed time:		0,134 sec.
Total compilation time:		0,029 sec.
**Total execution time:		0,105 sec.**
Number of compiled Spark inst:	1.
Number of executed Spark inst:	0.
Cache hits (Mem/Li/WB/FS/HDFS):	9/0/0/0/1.
Cache writes (Li/WB/FS/HDFS):	1/8/0/1.
Cache times (ACQr/m, RLS, EXP):	0,037/0,000/0,000/0,063 sec.
HOP DAGs recompiled (PRED, SB):	0/0.
HOP DAGs recompile time:	0,000 sec.
Spark ctx create time (lazy):	0,000 sec.
Spark trans counts (par,bc,col):0/0/0.
Spark trans times (par,bc,col):	0,000/0,000/0,000 secs.
Spark async. count (pf,bc,op): 	0/0/0.
Total JIT compile time:		36.052 sec.
Total JVM GC count:		176.
Total JVM GC time:		0.367 sec.
Heavy hitter instructions:
  #  Instruction  Time(s)  Count
  1  write          0,063      1
  2  sp_rblk        0,037      1
  **3  m_img_crop     0,005      1**
  4  -              0,002      4
  5  rmempty        0,001      1
  6  rshape         0,000      3
  7  rmvar          0,000      8
  8  +              0,000      8
  9  leftIndex      0,000      1
 10  rand           0,000      2

**image_translate**

SystemDS Statistics:
Total elapsed time:		0,166 sec.
Total compilation time:		0,049 sec.
**Total execution time:		0,117 sec.**
Number of compiled Spark inst:	1.
Number of executed Spark inst:	0.
Cache hits (Mem/Li/WB/FS/HDFS):	3/0/0/0/1.
Cache writes (Li/WB/FS/HDFS):	0/3/0/1.
Cache times (ACQr/m, RLS, EXP):	0,039/0,000/0,000/0,077 sec.
HOP DAGs recompiled (PRED, SB):	0/0.
HOP DAGs recompile time:	0,000 sec.
Spark ctx create time (lazy):	0,000 sec.
Spark trans counts (par,bc,col):0/0/0.
Spark trans times (par,bc,col):	0,000/0,000/0,000 secs.
Spark async. count (pf,bc,op): 	0/0/0.
Total JIT compile time:		13.584 sec.
Total JVM GC count:		176.
Total JVM GC time:		0.367 sec.
Heavy hitter instructions:
  #  Instruction      Time(s)  Count
  1  write              0,077      1
  2  sp_rblk            0,039      1
  **3  m_img_translate    0,001      1**
  4  leftIndex          0,000      1
  5  rmvar              0,000     12
  6  rand               0,000      1
  7  rightIndex         0,000      1
  8  createvar          0,000      5
  9  <                  0,000      8
 10  mvvar              0,000     10

Baunsgaard commented:
In the results you have above the time of the `Heavy hitter instructions:` is what i would use to compare. The rest of the time reported is all overhead of Java , I/O and other compilation related elements.
But i would modify the script you use. A example of a measuring script:

```R
res = read($1)
print(sum(res))
for(i in 1:$2) {
    res2 = OPERATION(res)
}
print(sum(res2))
```

This reads the file in the first argument and then repeats the operation the number of times specified in the second argument. Because of the way we execute in SystemDS the sum before the loop make sure res is actually read from disk.

j143 commented on src/test/java/org/apache/sysds/test/functions/nativ/ImgUtilsTest.java:
```diff
@@ -0,0 +1,521 @@
+        assertArrayEquals(expectedOutput, img_out, 1e-9); // Compare arrays with a small epsilon
+    }
+
+    public void runTests(String blasType) {
```
Hi @anishsapkota , there are a plenty of tests in the neighboring folders you can take inspiration from!

j143 commented:
Hi @anishsapkota -- let's move this forward, request the reviewer to commit this much by addressing the comments (which seem minor).


> could you please let me know which files I should be looking at, so that I can implement this as well ?

after that you can add further functionality.

Thanks, Janardhan

j143 commented:
Hi @anishsapkota -- thanks for your contribution. do you need some help here?

Baunsgaard commented:
> Hi @Baunsgaard , this PR does seems fine. only minor comments are there. Is it okay to take this in?

Hi @j143 , it can be taken in but we need to be careful, and go through the code, since some of the changes (for instance GitHub actions) contain changes that should not be made.

corepointer commented:
Hi!

Sorry that I'm a bit late to the discussion.

Our current Intel MKL support ends at version 2019.5 (you noticed the problems with dnn ops not compiling). I would be surprised if all magically worked by just using OneMKL and OneDNN.

We should probably put the native ops testing into a separate test suite and use a Docker container that supports it.

Furthermore, moving to OneAPI would change external behavior and require a major version bump.

Regards, Mark

j143 commented:
marking this PR for next release because there are some concerns pointed by @corepointer 

> Our current Intel MKL support ends at version 2019.5 (you noticed the problems with dnn ops not compiling). I would be surprised if all magically worked by just using OneMKL and OneDNN.
> 
> We should probably put the native ops testing into a separate test suite and use a Docker container that supports it.
> 
> Furthermore, moving to OneAPI would change external behavior and require a major version bump.
